src/databricks/labs/blueprint/paths.py

```diff
@@ -448,17 +441,296 @@ def match(self, path_pattern, *, case_sensitive=None):
+        self._ws.dbfs.mkdirs(self.as_posix())
+
+    def rmdir(self, recursive: bool = False) -> None:
+        """Remove a DBFS directory"""
+        self._ws.dbfs.delete(self.as_posix(), recursive=recursive)
+
+    def _rename(self: P, target: str | bytes | os.PathLike, overwrite: bool) -> P:
+        dst = self.with_segments(target)
+        if overwrite:
+            with dst.open(mode="wb") as writer, self.open(mode="rb") as reader:
+                shutil.copyfileobj(reader, writer, length=1024 * 1024)
+            self.unlink()
+        else:
+            self._ws.dbfs.move(self.as_posix(), dst.as_posix())
+        return dst
+
+    def unlink(self, missing_ok: bool = False) -> None:
+        """Remove a file in Databricks Workspace."""
+        if not missing_ok and not self.exists():
+            raise FileNotFoundError(f"{self.as_posix()} does not exist")
+        self._ws.dbfs.delete(self.as_posix())
+
+    def open(
+        self,
+        mode: str = "r",
+        buffering: int = -1,
+        encoding: str | None = None,
+        errors: str | None = None,
+        newline: str | None = None,
+    ):
+        """Open a DBFS file.
+
+        Only text and binary I/O are supported in basic read or write mode, along with 'x' to avoid overwriting."""
+        is_write = "w" in mode
+        is_read = "r" in mode or not is_write
+        if is_read and is_write:
+            msg = f"Unsupported mode: {mode} (simultaneous read and write)"
+            raise ValueError(msg)
+        is_binary = "b" in mode
+        is_text = "t" in mode or not is_binary
+        if is_binary and is_text:
+            msg = f"Unsupported mode: {mode} (binary and text I/O)"
+            raise ValueError(msg)
+        is_overwrite = is_write and "x" not in mode
+        binary_io = self._ws.dbfs.open(self.as_posix(), read=is_read, write=is_write, overwrite=is_overwrite)
+        if is_text:
+            return io.TextIOWrapper(binary_io, encoding=encoding, errors=errors, newline=newline)
+        return binary_io
+
+    def write_bytes(self, data):
+        """Write the (binary) data to this path."""
+        # The DBFS BinaryIO implementation only accepts bytes and rejects the (other) byte-like builtins.
+        match data:
+            case builtins.bytes | builtins.bytearray:
+                binary_data = bytes(data)
+            case _:
+                binary_data = bytes(memoryview(data))
+        with self.open("wb") as f:
+            return f.write(binary_data)
+
+    @property
+    def _file_info(self) -> FileInfo:
+        # this method is cached because it is used in multiple is_* methods.
+        # DO NOT use this method in methods, where fresh result is required.
+        try:
+            return self._cached_file_info
+        except AttributeError:
+            self._cached_file_info = self._ws.dbfs.get_status(self.as_posix())
+            return self._cached_file_info
```

Comment:
```suggestion
    @cached_property
    def _file_info(self) -> FileInfo:
        # this method is cached because it is used in multiple is_* methods.
        # DO NOT use this method in methods, where fresh result is required.
        return self._ws.dbfs.get_status(self.as_posix())
```
and in factory - `path.__dict__['_file_info'] = file_info`

Reply:
These classes can't currently use `@cached_property` because it [doesn't support `__slots__`](https://docs.python.org/3/library/functools.html).

These classes started off slot-based because the superclasses also were. Although I could switch to dict-based attributes, I'd prefer to keep these slot-based: I've normally always preferred slots anyway because: a) errors are immediate, including during `__init__`; b) IDE support is better; c) they're more efficient.

Do you still want me to make this change?

Reply:
add `threading.RLock()` for refreshing file info then, as that's what cached property does as well

Reply:
It looks like I missed the merge deadline for this. That said, [in 3.12 they've removed that lock](https://docs.python.org/3/library/functools.html#functools.cached_property):

> Changed in version 3.12: Prior to Python 3.12, cached_property included an undocumented lock to ensure that in multi-threaded usage the getter function was guaranteed to run only once per instance. However, the lock was per-property, not per-instance, which could result in unacceptably high lock contention. In Python 3.12+ this locking is removed.